Sense for VirtualBox and $HOME when deciding to turn on vagrant testing.

jbaiera opened the pull request:
Some of the CI boxes for Elasticsearch have inconsistently configured environments when it comes to Vagrant. Vagrant requires $HOME to be set, as well as VirtualBox to be installed. This PR checks those two things are present before enabling tests with VagrantFixtures in the HDFS Repo (only place vagrant fixtures are used right now).

jbaiera committed:
Now sensing for virtualbox as well as the HOME environment variable when deciding if vagrant is supported

rjernst commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -154,29 +154,70 @@ if (fixtureSupported) {
+
+  // Finding that HOME needs to be set when performing vagrant updates
+  String homeLocation = System.getenv("HOME")
+  if (homeLocation == null) {
```
why do we need this leniency? If vagrant fails because HOME is not set, that tails us there is a problem that needs fixing. With the logic here, we will simply not run the test and not know it is broken.

jbaiera replied:
I can change this to throw an exception instead. If vagrant is installed but the environment can't support calling it, we should probably break earlier in the build with clear reasoning for why.

rjernst commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -154,29 +154,70 @@ if (fixtureSupported) {
+  try {
+    ByteArrayOutputStream pipe = new ByteArrayOutputStream()
+    ExecResult runResult = exec {
+      commandLine 'vboxmanage', '--version'
```
We really need to share this stuff...VagrantTestPlugin also does this.

Can we do this exactly once (for vagrant and vboxmanage each) and set a property on the project like `vagrantSupported`?  See for example in BuildPlugin how we check the java version etc in globalBuildInfo. This would also allow us to remove the tasks created by VagrantTestPlugin and just check the flag.

dimitris-athanasiou commented:
@jbaiera I have merged in https://github.com/elastic/elasticsearch/pull/24643 in order to temporarily disable the task that causes build failures. Adding it here to remind you to re-enable once this is ready to go in.

jbaiera committed:
Add a plugin for determining if a build environment supports Vagrant on VirtualBox.

jbaiera committed:
Some fixes to a few bugs in the new files

jbaiera committed:
Actually call verify

jbaiera committed:
move previous vagrant based projects to use support plugin.

jbaiera committed:
simplify repo hdfs build script.

jbaiera committed:
re-enable secure testing in repo hdfs

jbaiera commented:
I've moved the environment sensing out of the hdfs plugin build script as well as the vagrant test plugin into it's own `VagrantSupportPlugin`. Both vagrant and virtualbox are checked to see if they are installed. After the plugin is applied, if the root project hasn't been configured by it yet, then it does the environment checks and installs the results (as `Installation` objects) into the root project. It then transfers the `Installation` objects into the project applying the plugin. `Installation` is just a simple object that couples together the results of the checks: If the commands are installed, if the versions are correct, what the version number is if they meet the requirements, and any errors encountered while checking if they were installed. The plugin also creates the tasks to check the version for Vagrant and VirtualBox, so that hdfs repo and the packaging tests don't have to any more. These tasks no longer perform the actual lookup, but rather call the `verify()` method on the Installation objects, which throws if the Installation is not valid.

jbaiera commented:
The VagrantTestPlugin definitely depends on this VagrantSupportPlugin. I'm not aware of any good ways to orchestrate that dependency in the build script. @rjernst any thoughts?

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantSupportPlugin.groovy:
```diff
@@ -0,0 +1,133 @@
+    }
+
+    private Installation getVirtualBoxInstallation(Project project) {
+        // Also check to see if virtualbox is installed. We need both vagrant
```
This comment also refers to stuff outside this plugin

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/Installation.groovy:
```diff
@@ -0,0 +1,47 @@
+ * is installed, if it is the correct version, which version was found, and any
+ * errors that were encountered when checking for the installation.
+ */
+class Installation {
```
I don't think we really need this abstraction...can we just keep it simple and store a boolean property on the root project indicating whether vagrant is installed at the given version, and the same for virtualbox? You can even store this in a map, along with whatever error description you want about getting the version. eg.

```
rootProject.ext.vagrant = ['supported': false, 'info': 'Unsupported version of vagrant [${version}]. Need [Vagrant 1.8.6+]']
....
if (rootProject.vagrant.supported == false) {
    logger.warn(rootProject.vagrant.info)
} else {
    // use vagrant
}
```

rjernst commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -156,31 +159,21 @@ if (fixtureSupported) {
-    } else {
-      logger.warn("Could not read installed vagrant version:\n" + output)
+    if (project.virtualBoxInstallation.supported == false) {
+      logger.warn(project.virtualBoxInstallation.error.message)
```
These messages are becoming noisy, and aren't actually helpful, because they are emitted even when not trying to run this task. I wonder if instead you could check emit this at runtime if the task would have run? You could do this as a doFirst, and throw a StopExecutionException if the task should be skipped.

rjernst reviewed (commented):
I left a few comments.

jbaiera committed:
Addressing comments from PR:

Removing Installation and replacing with Map
Removing placing the installation properties on the project
Got rid of spammy log messages on the repository hdfs build

jbaiera commented:
@rjernst How are we on the recent changes for this? I'd like to get this merged soon.

rjernst commented:
@jbaiera I had not looked at it since I left my last comments. Please ping me in the future when it is ready for another review. I will look now.

rjernst commented on buildSrc/src/main/groovy/org/elasticsearch/gradle/vagrant/VagrantSupportPlugin.groovy:
```diff
@@ -0,0 +1,127 @@
+
+        // Finding that HOME needs to be set when performing vagrant updates
+        String homeLocation = System.getenv("HOME")
+        if (project.rootProject.ext.vagrantSupported && homeLocation == null) {
```
Why does this check need to be done for every project using this instead of once inside the installation checks above?

rjernst commented on plugins/repository-hdfs/build.gradle:
```diff
@@ -153,60 +156,39 @@ if (fixtureSupported) {
+}
 
+if (secureFixtureSupported) {
   integTestSecure.mustRunAfter(project.integTest)
```
I don't think this mustRunAfter will actually do anything. The integTest task is just a dummy task that depends on setting up the integTestCluster and running integTestRunner. When you run gradle check do you see tasks form the secure and regular integTest intermixed?

rjernst reviewed (approved):
LGTM, left a couple minor comments that can be addressed in followups if desired.

jbaiera committed:
Hoisting the home check into the single execution only

jbaiera committed:
removing the run after stipulation as it is not needed
